### PR TITLE
Codegen multistate

### LINKED ISSUE #157: Codegen does not allow neighbourhoods to be be referenced in functions
In code gen neighbourhoods are constructed by calling a function. That function requires a world view, and that world view is a local variable passed to the function that evaluates the state logic. So if the state logics calls a custom or builtin function, the world view is no longer available.

We can solve this by making each local neighbourhoods to global variables.

### PR DESCRIPTION
**Pull request by jenrik**
Adds support for multistates in the Kotlin code generator. Also fixes functions not having any arguments, and several unreachable visitor functions having a TODO statement instead of a proper error.

This pull request depends on changes in #158 
Resolves #157

### CODE REVIEW

File: compiler/src/main/kotlin/visitors/codegen.kt [hunk 1]
```diff
@@ -169,6 +183,87 @@ class KotlinCodegen : ASTVisitor<String> {
...
+    }
+
+    private fun emitMultiStateLookup(states: List<StateDecl>): String {
+        val builder = StringBuilder()
+
+        builder.append("""
+            fun baseMultiStateLookup(state: Int, index: Int): Int {
+
+            if (index < 0) {
+            ${INDENT}throw Error()
```
Comment (NicEastvillage) [inline compiler/src/main/kotlin/visitors/codegen.kt]: Maybe use OutOfBoundsException?

File: compiler/src/main/kotlin/visitors/codegen.kt [hunk 2]
```diff
@@ -169,6 +183,87 @@ class KotlinCodegen : ASTVisitor<String> {
...
+            val baseId = baseStateLookup(state)
+            ${INDENT}return when(baseId) {
+        """.trimIndent())
+
+        states.forEach {
+            val baseId = stateIDs[it.ident]!!
+            builder.append("""
+                $INDENT$INDENT$baseId -> {
+                $INDENT$INDENT${INDENT}if(index >= ${it.multiStateCount}) {
+                $INDENT$INDENT$INDENT${INDENT}throw Error()
```
Comment (NicEastvillage) [inline compiler/src/main/kotlin/visitors/codegen.kt]: Maybe use OutOfBoundsException?

File: compiler/src/main/kotlin/visitors/codegen.kt [hunk 3]
```diff
@@ -558,19 +656,39 @@ class KotlinCodegen : ASTVisitor<String> {
...
 
     override fun visit(node: FuncCallExpr): String {
-        return "(${getMappedLabel(node.ident)}())"
+        val builder = StringBuilder()
+        builder.append("(")
+        builder.append(getMappedLabel(node.ident))
+        builder.append("(worldView")
+        node.args.forEachIndexed { index, expr ->
+            if (index > 0) {
+                builder.append(", ")
```
Comment (NicEastvillage) [inline compiler/src/main/kotlin/visitors/codegen.kt]: Are you sure you got a comma between worldView and first argument?

**Review by NicEastvillage — CHANGES_REQUESTED** [review]

**Comment by NicEastvillage** [comment]
Your PR does not resolve #157 because FuncCall is now broken for built-in functions

**Comment by jenrik** [comment]
Builtins have been updated to reflect the changes to the function argument convention

**Review by NicEastvillage — APPROVED** [review]
Nice

**Pull request merged**